### PR TITLE
Fix typehint for iterables.

### PR DESCRIPTION
**Pull request by ADmad**
Improve docblock types to help static analyzers and IDEs detect the type of item returned by iterables.

https://blog.jetbrains.com/phpstorm/2016/06/phpstorm-2016-2-eap-162-844/

### CODE REVIEW

**Comment by dereuromark** [comment]
How does this solve the inline annotation needed?
```php
/** @var \Cake\ORM\Association[] $associations */
$associations = $modelInstance->associations();
foreach ($associations as $association) {
	$relationType = $association->type();
```

**Comment by othercorey** [comment]
Wonder when we can officially use the class-string type I added to TypeResolver.

File: src/Console/CommandCollection.php [hunk 1]
```diff
@@ -34,6 +35,7 @@ class CommandCollection implements IteratorAggregate, Countable
      * Command list
      *
      * @var array
```
Comment (othercorey) [inline src/Console/CommandCollection.php]:
```suggestion
     * @var (\Cake\Console\Shell|\Cake\Console\CommandInterface|string)[]
```
Can we at least do this?

File: src/Core/ObjectRegistry.php [hunk 2]
```diff
@@ -375,10 +376,10 @@ public function unload(string $objectName)
     /**
      * Returns an array iterator.
      *
-     * @return \ArrayIterator
-     * @psalm-return \ArrayIterator<string, TObject>
+     * @return \Traversable
```
Comment (othercorey) [inline src/Core/ObjectRegistry.php]:
```suggestion
     * @return \Traversable<string, TObject>
```
TypeResolver should able to handle this. It generates a Collection type.

**Comment by ADmad** [comment]
> How does this solve the inline annotation needed?

Did you read type hinting section on https://blog.jetbrains.com/phpstorm/2016/06/phpstorm-2016-2-eap-162-844/?  It's shows an example for exactly the usage you show.

**Comment by ADmad** [comment]
> Wonder when we can officially use the class-string type I added to TypeResolver.

Probably not until phpstorm starts supporting it, same for the other changes you have suggested. Everyone's beloved IDE is what's preventing us from using better annotation types.

**Comment by othercorey** [comment]
> Probably not until phpstorm starts supporting it, same for the other changes you have suggested. Everyone's beloved IDE is what's preventing us from using better annotation types.

phpstorm supports compound array types at least, right?

**Comment by ADmad** [comment]
> phpstorm supports compound array types at least, right?

IDK, if someone confirms I can make the changes.

**Comment by dereuromark** [comment]
I can confirm that it works in PHPStorm 2019.3.4 etc
@ADmad thank you for looking into this.

**Comment by dereuromark** [comment]
@othercorey If you want me to test something else, please make a branch and give that to me.

**Pull request merged**